### PR TITLE
Add common lombok config

### PR DESCRIPTION
**Pull request by Tristan-WorkGH**
**Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines
- [x] ~Tests for the changes have been added (for bug fixes / features)~
- [x] Docs have been added / updated (for bug fixes / features)

**Does this PR already have an issue describing the problem?**
No.

**What kind of change does this PR introduce?**
Add a profile (inactive by default) to be use by children projects by activating the correct profile.  
It consist of:
  * the common base of lombok configuration in the `/lombok.config` to be reuse by children
  * a check that the configuration is correctly referenced in the project using it

**What is the current behavior?**
Each projects have their own `lombok.config` with duplication and often incomplete.

**What is the new behavior (if this is a feature change)?**
Each children project can have a common base and update can be centralized.

**Does this PR introduce a breaking change or deprecate an API?**
No.

**Other information**:
The idea is to use lombok's [`import`](https://projectlombok.org/features/configuration#import) feature to have a common base for GridSuite projects.

Projects that want to use this project will need to:
  * have this module as parent
  * activate this profile: _(one of the following)_
    - using command line with `-Plombok-config-copy` and `-Plombok-config-check`
    - create the file `.mvn/lombok-config-copy` in the project to activate the copy of the file to be used
    - optionally create the file `.mvn/lombok-config-check` in the project to activate the check that `lombok.config` contains the import
  * add this line _**at the start**_ of `/lombok.config`
      ```properties
      import target/configs/powsybl-build-tools.jar!powsybl-build-tools/lombok.config
      ```

This profiles will, when activate:
  1. during `initialize` phase: copy the *powsybl-build-tools* jar in `target/configs/` (to be found and use by lombok)
  2. during `validate` phase: check that the `lombok.config` file have the correct import at the start of the file

This has been tested with Maven, JetBrains IDEA IntelliJ and Eclipse (_with m2e + lombok plugin_).

The `initialize` has been use because of its early position in (Maven lifecycle)[https://maven.apache.org/guides/introduction/introduction-to-the-lifecycle.html], as IDEs seems to run this phase when reloading/refreshing the project and we need the jar to be present for IDE to parse the configuration (when editing files).

_**Side note**:
No better way found, mainly because of limitation in the way lombok manage internally the import path ([ConfigurationFile](https://github.com/projectlombok/lombok/blob/9e5e0c3803dd5e4306bb54b4976d1c03ec1a8ee2/src/core/lombok/core/configuration/ConfigurationFile.java#L124) and [FileSystemSourceCache](https://github.com/projectlombok/lombok/blob/9e5e0c3803dd5e4306bb54b4976d1c03ec1a8ee2/src/core/lombok/core/configuration/FileSystemSourceCache.java#L62)).
And [that will not change in near future](//github.com/projectlombok/lombok/issues/2158)._

### CODE REVIEW

File: powsybl-build-tools/src/main/resources/powsybl-build-tools/lombok.config [hunk 1]
```diff
@@ -0,0 +1,14 @@
+# Don't search another lombok.config file in parent folders
+config.stopbubbling = true
+
+# We add annotations flags for IDEs and tools :
+#  * mark the code as generated for linters and coverage tools
+lombok.addJavaxGeneratedAnnotation = true
```
Comment (jonenst) [inline powsybl-build-tools/src/main/resources/powsybl-build-tools/lombok.config:6]: Do we want this ? I see in the docs:
" BREAKING CHANGE: Starting with lombok v1.16.20, defaults to false instead of true, as this annotation is broken in JDK9."

File: powsybl-parent/pom.xml [hunk 2]
```diff
@@ -352,5 +355,73 @@
...
+                                    <goal>copy</goal>
+                                </goals>
+                                <configuration>
+                                    <artifactItems>
+                                        <artifactItem>
+                                            <groupId>com.powsybl</groupId>
+                                            <artifactId>powsybl-build-tools</artifactId>
+                                            <version>${powsybl-build-tools.version}</version>
+                                            <type>jar</type>
+                                            <overWrite>false</overWrite>
```
Comment (jonenst) [inline powsybl-parent/pom.xml:389]: should be true ?

File: powsybl-parent/pom.xml [hunk 3]
```diff
@@ -352,5 +355,73 @@
...
+                                <id>lombok-verify-main-file</id>
+                                <phase>process-sources</phase>
+                                <goals>
+                                    <goal>enforce</goal>
+                                </goals>
+                                <configuration>
+                                    <rules>
+                                        <evaluateBeanshell>
+                                            <condition>org.codehaus.plexus.util.FileUtils.fileRead(new File("lombok.config"), "UTF-8").startWith("import target/configs/powsybl-build-tools.jar!powsybl-build-tools/lombok.config")</condition>
+                                            <!--<condition>java.nio.file.Files.lines(java.nio.file.Paths(".", "lombok.config")).anyMatch(s -> "import ./target/configs/powsybl-build-tools.jar!powsybl-build-tools/lombok.config".equals(s))</condition>-->
```
Comment (jonenst) [inline powsybl-parent/pom.xml:415]: remove commented code or state the reason why it is kept in a comment

File: powsybl-parent/pom.xml [hunk 4]
```diff
@@ -352,5 +355,73 @@
...
         </profile>
+        <profile>
+            <id>lombok-config</id>
+            <!-- Add at the start of lombok.config this line: import target/configs/powsybl-build-tools.jar!powsybl-build-tools/lombok.config -->
+            <activation>
+                <property>
+                    <name>powsybl.lombok.config</name>
+                    <value>true</value>
+                </property>
+                <!--<file>
```
Comment (jonenst) [inline powsybl-parent/pom.xml:366]: remove commented code

File: powsybl-build-tools/src/main/resources/powsybl-build-tools/lombok.config [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+lombok.addLombokGeneratedAnnotation = true
+#  * disable all warnings on generated code for checking tools and IDE (eg. checkstyle)
+lombok.addSuppressWarnings = true
```
Comment (jonenst) [inline powsybl-build-tools/src/main/resources/powsybl-build-tools/lombok.config:9]: Remove ? From the docs
"Lombok adds the @SuppressWarnings("all") annotation to all generated nodes by default."

File: powsybl-build-tools/src/main/resources/powsybl-build-tools/lombok.config [hunk 6]
```diff
@@ -0,0 +1,14 @@
...
+#  * we prefer the javax Nullable annotation for compatibility with majority of tools
+lombok.addNullAnnotations = javax
```
Comment (jonenst) [inline powsybl-build-tools/src/main/resources/powsybl-build-tools/lombok.config:11]: "Many flavors are available: javax (=JSR305; not recommended), "
why do we use a not recommended conf ?

Comment (jonenst) [reply powsybl-build-tools/src/main/resources/powsybl-build-tools/lombok.config:11]: remove ?

File: powsybl-parent/pom.xml [hunk 7]
```diff
@@ -352,5 +355,73 @@
...
+                                    <outputDirectory>${project.build.directory}/deps</outputDirectory>
+                                    <overWriteReleases>false</overWriteReleases>
+                                    <overWriteSnapshots>true</overWriteSnapshots>
+                                </configuration>
+                            </execution>
+                        </executions>
+                    </plugin>
+                    <plugin>
+                        <groupId>org.apache.maven.plugins</groupId>
+                        <artifactId>maven-enforcer-plugin</artifactId>
```
Comment (jonenst) [inline powsybl-parent/pom.xml:403]: can we have this in the "checks" profile/property so that it may be turned off like other checks?

**Comment by Tristan-WorkGH** [comment]
Due to limitation in Maven profiles activation system, like in #2, we use a system of file markers to activate a profile from child projects.

Another note is during my tests, it seems we can't have multiple conditions like `<file/>` + `<property/>` or `<file><exists/><exist/></file>`, as maven only check the first condition...  
So the following _don't work_ for activate if a file exist and another not exist at same time: `<file><exists>file1</exists><missing>file2</missing></file>`

File: powsybl-parent/pom.xml (hunk 7 above)
Comment (Tristan-WorkGH) [reply powsybl-parent/pom.xml:412]: done in commit with marker files

File: powsybl-build-tools/src/main/resources/powsybl-build-tools/lombok.config (hunk 1 above)
Comment (Tristan-WorkGH) [reply powsybl-build-tools/src/main/resources/powsybl-build-tools/lombok.config:6]: I think it's with jigsaw/modules system that it broke, but we don't use it. _See (lombok#985)[https://github.com/projectlombok/lombok/issues/985]. _  
After did we consider that some projects use/will use it?

File: powsybl-build-tools/src/main/resources/powsybl-build-tools/lombok.config (hunk 6 above)
Comment (Tristan-WorkGH) [reply powsybl-build-tools/src/main/resources/powsybl-build-tools/lombok.config:11]: > "Further more these annotations use javax. packages which it is not possible to use according to the Oracle Java binary licence [...]"
Oh, I haven't check that... I switch to this one at beginning because the majority of tools recognise javax annotations natively.  
Between the possibilities (~`javax`~, `jakarta`, `eclipse`, `jetbrains`, `netbeans`, ~`androidx`~, ~`android.support`~, `checkerframework` _(recommended)_, `findbugs`, `spring`, `jml`), I wanted to use `checkerframework`/`findbugs`/`spring`, but there aren't in dependencies by default ...

File: powsybl-build-tools/src/main/resources/powsybl-build-tools/lombok.config (hunk 5 above)
Comment (Tristan-WorkGH) [reply powsybl-build-tools/src/main/resources/powsybl-build-tools/lombok.config:9]: hum, yes, i remove it

File: powsybl-parent/pom.xml [hunk 8]
```diff
@@ -352,5 +355,98 @@
...
+                                    <goal>copy</goal>
+                                </goals>
+                                <configuration>
+                                    <artifactItems>
+                                        <artifactItem>
+                                            <groupId>com.powsybl</groupId>
+                                            <artifactId>powsybl-build-tools</artifactId>
+                                            <version>${powsybl-build-tools.version}</version>
+                                            <type>jar</type>
+                                            <overWrite>true</overWrite>
```
Comment (jonenst) [inline powsybl-parent/pom.xml:386]: > Using the default settings (overWriteReleases = false, overWriteSnapshots = false, overWriteIfNewer = true), then a release or snapshot artifact will only over write the destination if the source is newer than the destination (or marker file if unpacking).

Why don't we use the default

File: powsybl-parent/pom.xml [hunk 9]
```diff
@@ -352,5 +355,98 @@
...
+                                            <outputDirectory>${project.build.directory}/configs</outputDirectory>
+                                            <destFileName>powsybl-build-tools.jar</destFileName>
+                                        </artifactItem>
+                                    </artifactItems>
+                                    <outputDirectory>${project.build.directory}/deps</outputDirectory>
+                                    <overWriteReleases>false</overWriteReleases>
+                                    <overWriteSnapshots>true</overWriteSnapshots>
```
Comment (jonenst) [inline powsybl-parent/pom.xml:393]: why don't we use the default

File: powsybl-parent/pom.xml [hunk 10]
```diff
@@ -352,5 +355,98 @@
...
+                                </configuration>
+                            </execution>
+                        </executions>
+                    </plugin>
+                    <plugin>
+                        <artifactId>maven-compiler-plugin</artifactId>
+                        <configuration>
+                            <showWarnings>true</showWarnings><!--show lombok warning during compile-->
```
Comment (jonenst) [inline powsybl-parent/pom.xml:401]: separate PR

File: powsybl-parent/pom.xml [hunk 11]
```diff
@@ -352,5 +355,98 @@
...
+                                <id>lombok-verify-main-file</id>
+                                <phase>validate</phase>
+                                <goals>
+                                    <goal>enforce</goal>
+                                </goals>
+                                <configuration>
+                                    <rules>
+                                        <requireActiveProfile>
+                                            <profiles>lombok-config-copy</profiles>
+                                            <message>The profile "lombok-config-copy" isn't active, please use '-P...' or file marker to activate it.</message>
```
Comment (jonenst) [inline powsybl-parent/pom.xml:430]: remove -P message. Only document using the marker

File: powsybl-parent/pom.xml [hunk 12]
```diff
@@ -257,7 +260,7 @@
                           <dependency>
                             <groupId>com.powsybl</groupId>
                             <artifactId>powsybl-build-tools</artifactId>
-                            <version>12-SNAPSHOT</version>
+                            <version>${powsybl-build-tools.version}</version>
```
Comment (jonenst) [inline powsybl-parent/pom.xml:263]: why this change ?

**Comment by Tristan-WorkGH** [comment]
branch updated

**Review by jonenst — APPROVED** [review]
